### DISCUSSION TITLE
Sending Access-Control-Expose-Headers response header

### DISCUSSION THREAD

**Discussion by tlinhart** [snok/asgi-correlation-id#66] [Ideas]
Hi, setting the `Access-Control-Expose-Headers` response header with this middleware feels a bit wrong. There a couple of reasons.

At the moment the header is always set, even for non-CORS requests. This doesn't cause any harm but feels superfluous. It should probably be used only in context of a [CORS protocol](https://fetch.spec.whatwg.org/#http-cors-protocol).

If you want to allow propagation of correlation ID from the downstream services/components, you have to handle CORS in its entirety, especially responding with `Access-Control-Allow-Headers` to preflight requests to actually allow sending request headers with correlation ID set. This is typically done with a specialized middleware like Starlette's [CORS middleware](https://www.starlette.io/middleware/#corsmiddleware) which has its own means of setting the expose headers. If you use both middlewares together the `Access-Control-Expose-Headers` header is duplicated in the response. This again is strictly not a problem (according to [RFC 9110](https://www.rfc-editor.org/rfc/rfc9110.html#name-field-order)) but again feels superfluous especially when you need to expose multiple headers and use CORS middleware to handle this.

What do you think @JonasKs @sondrelg ?

**Comment by sondrelg** [comment]
You're suggesting we just drop that part entirely? 

I agree, it seems like specifying `expose_headers` in the cors middleware is the way to go if you want these headers attached to your response. For context, this middleware basically inherited that piece of logic from [here](https://github.com/snok/django-guid/pull/24).

This feels like major bump territory, right?

**Reply by tlinhart** [reply]
That's basically what I wanted to say :-)

I'm convinced it's a move in a right direction that would lead to a more compact package with a cleaner scope and better separation of concerns. Exposing the correlation ID header via this middleware alone is only useful in the case when there are no downstream services or components that would send the correlation ID in the request headers. Only in this case you are fine with just exposing the response header and not taking care of also allowing request headers. But this is rarely the real world setup. You typically want the request ID to be created at the top level service or component (e.g. frontend) so you can track the request throughout the whole stack. In that case, you need to handle CORS holistically. So the current state is more like a notice about something that needs to be taken care of than a complete solution to the problem. In that case, it may be a better idea to get rid of the partial solution in favor of presenting the complete and universal solution in the documentation.

I'm aware this might look like a major change but for the reasons presented above, I'm quite confident it wouldn't break a lot in practise.

**Reply by sondrelg** [reply]
I don't mind bumping major versions for this. It's a backwards incompatible change, so seems appropriate :+1: 

I guess it might be enough to just drop the code for now. We could document CORS-stuff, but seems out of scope for this package. Might reconsider if people report issues related to this.

Anything to add to this @tlinhart @JonasKs?

I'll let you create the PR of you want to @tlinhart - otherwise I'd be happy to do it later today or tomorrow :slightly_smiling_face:

**Reply by tlinhart** [reply]
I agree with both points. I'll prepare a PR referring this discussion. Thanks!

**Reply by JonasKs** [reply]
I think documenting it is kind of essential to the ease of use, and therefor adoption of the package. 

I don’t mind the change though, but I am strongly in favor of documentation on how to do it correct.

**Reply by tlinhart** [reply]
I can try to add a documentation proposal (README update) if you agree.

**Reply by sondrelg** [reply]
Yeah sure :clap: Something like a "if you're using this with CORS, you'll need to expose these headers.." kind of section? I don't really remember what your original use-case was @JonasKs, would you mind explaining it if you still remember?

**Reply by JonasKs** [reply]
Sorry for my somewhat abrupt response before, was on my way out and wanted to reply quickly. 

> it may be a better idea to get rid of the partial solution in favor of presenting the complete and universal solution in the documentation.

Strongly for. 

> I'm quite confident it wouldn't break a lot in practise.

Any breakage = major, but we're not afraid to major bump here. Move quick, break things for the right reasons 😊

> "if you're using this with CORS, you'll need to expose these headers.."

Yeah, exactly!  
I think the real use case is that if you don't explicitly say these headers are to be returned, and the browser has implemented CORS, you won't be able to access the header. In other words, your javascript frontend won't be able to display the correlation ID to the end user. 😊

**Reply by tlinhart** [reply]
OK, thank you. I'll try to prepare something along these lines tomorrow.

**Reply by tlinhart** [reply]
@JonasKs @sondrelg 

I've just updated the README, could you please take a look if it's complete from your point of view?

**Reply by sondrelg** [reply]
Looks great.

One final point of discussion before we release this. Since we're bumping major versions anyways, perhaps it would now make sense to change the behaviour of the newly added `update_request_header` parameter. Seems like defaulting to `True` would be most useful.

What do you think?

**Reply by tlinhart** [reply]
It's a different concept but I agree that this is a good opportunity to introduce the change. So I'm for if you ask me :-)

**Reply by JonasKs** [reply]
Agree!

**Comment by tlinhart** [comment]
@sondrelg
I created a PR (https://github.com/snok/asgi-correlation-id/pull/67) but it [fails the Python 3.7 test](https://github.com/snok/asgi-correlation-id/actions/runs/4233749664/jobs/7355109522). Personally I doubt it's related to the changes and more likely relates to https://github.com/snok/asgi-correlation-id/commit/5a7be6337f3b33b84a00d03baae3da999bb722d5 given the `DeprecationWarning` in the workflow log. What do you think?

**Reply by tlinhart** [reply]
Yeah, I guess it's a problem with `importlib-metadata` package. If I compare the failing workflow run with the [yesterday's run](https://github.com/snok/asgi-correlation-id/actions/runs/4227554108/jobs/7342140132) which passes, different versions of the package are being installed.

**Reply by sondrelg** [reply]
Ah ok, don't worry about it :+1: I'll take a look later and fix it separately if needed
